### ISSUE TITLE
Package-wide analysis

### ISSUE THREAD

**Issue by chavacava** [mgechev/revive#46]
Hi @mgechev, I would love to develop some rules that need package-wide analysis (e.g. `unused-function`) but I do not see how to implement the analysis pattern required by these kind of rules.

I've developed one package-wide rule (`confusing-naming`). A raw description of the rule is: create a global registry of names, and for each linted file: a) enrich that list, and b) check new names vs those in the list.
Because two names are similar (_confusing_) no matter the order in which we find them, checks are independent of: the order in which package files are analyzed, and the global progress of the analysis.

That is not the case of, for example, `unused-function`: checks can be done only when all files in the package were analyzed (it is not possible to flag a function as `unused` **until** all files of the package have been analyzed). These kind of rules need something like a _reduce_ phase after been applied to all package's files.

My understanding is that rules are applied as follows:

```
for each file in the package
  go { 
    for each rule
      rule.Apply(file)
    send failures
  }
```

Files are concurrently analyzed (and that is good :+1: ), thus at rule-level it is impossible to have information of the global (package-level) progress of the analysis.

Is my understanding  wrong? Do you see how to implement a rule like  `unused-function` without modifying the linter logic?

**Comment by mgechev** [comment]
Yes, you're correct. I see three potential solutions to this issue:
1. Allow rules to request a sequential application. This is similar to the optimization we do for rules which require type checking.
2. Collect data with "meta-rules" which run in parallel and when all meta-rules for the entire package are done collecting data, invoke package-wide rules.

I like 2.; the meta-rules can be the same as all other rules, we'll not document them and disallow publicly using them. They won't report anything; they'll collect data and return an object, which the linter will pass to the package-wide rules later on.

**Comment by chavacava** [comment]
As you, I think the second approach is better than the first.
A third approach could be:

* Let package-wide rules implement a complementary interface (say `Reducer`, yes, the name is not terrific but...). That interface requires to implement the `Reduce` method that takes a `lint.Package` and yields a list of package level failures.
* In `package.lint` method, after all files were linted (i.e. after the `wg.Wait()`), iterate over the rules and for all those that implement the `Reducer` interface, call the `Reduce` method to retrieve package-level failures.

The approach has the advantage that there is no need to introduce a new _kind_ of rules, in fact a rule can generate both file-level and package-level failures.

I've made a proof-of-concept of the approach, you can check-it in PR #48 

What do you think?

**Comment by mgechev** [comment]
This sounds good. What would be the `Apply` method of `Reducer` rules? In the general case, it looks like they'll just return? In such case, maybe it's worth having rules which can implement either the `Rule` or `Reducer` (maybe `Reducer` -> `PackageRule`, `Rule` -> `FileRule`) interface?

**Comment by chavacava** [comment]
> What would be the Apply method of Reducer rules?
> In the general case, it looks like they'll just return?

Package-level rules **still need** `Apply` method to collect data at file level. In the general case the **result** of `Apply` is an empty slice.(1)

For example, the rule to detect unused functions could be implemented as follows:
1. Use `Apply` method collect data on functions defined and used at file-level (in this case `Apply` will call, as standard rules do, `ast.Walk` and visit the AST of the file looking for func decls and calls
2. Use the `Reduce` method to analyse collected data and determine if there are  unused functions.

(1) a rule may, in exceptional cases, generate both kind of failures file-level and package level

**Comment by chavacava** [comment]
Yes, concurrency-safety is a concern (in #48, at `Apply`, I need to use a mutex to avoid race conditions on the map where data is collected)

Do we  delegate  to the `Apply(p *Package)` method the responsibility of iterating over all package files? 

> _This way we also have the guarantee that they'll be thread safe (i.e., one rule instance per package)_

:warning: there is no  guarantee that `Apply(p *Package)` will iterate over files in a thread-save manner.

**Comment by mgechev** [comment]
👍it's a responsibility to the consumer to guarantee this.

**Comment by chavacava** [comment]
Wrapping-up my understanding of the proposed solution:

1. Add a new interface 
```go
type PackageRule interface {
  Apply(*lint.Package) []PkgFailure
  Name() string
}
```
2. At `Package.lint` iterate over rules and call `Apply` of those implementing the `PackageRule` interface and maintain current behaviour (iterate over files and call `file.lint`)
At the end of execution of an `Apply(pkg)` handle failures like `File.lint` do.

I'm ok?

**Comment by mgechev** [comment]
This sounds good to me.

**Comment by chavacava** [comment]
How do you want to proceed? Do you go ahead with the modification or do I take it?

**Comment by mgechev** [comment]
In the next ~3 weeks, I'll not be able to make any significant work on open source. You can go ahead when you have some time.

**Comment by chavacava** [comment]
:+1:

**Comment by mgechev** [comment]
@chavacava let's close this one for now. There are not too many rules which require package-wide analysis, so it might not be worth it to introduce additional complexity or slow the execution down.

**Comment by chavacava** [comment]
I'm in the middle of the implementation for this issue. The approach, due to modifications on function signatures, requires some deep refactoring.

Ok to put it in standby

**Comment by ademenev** [comment]
Any news on this? I would really love to see support for package-wide rules in revive!

**Comment by chavacava** [comment]
@ademenev do you have some rules in mind?
I imagine rules like: unused private function/struct/field

**Comment by ademenev** [comment]
Unused unexported symbols is what I want mostly
